rfd/0180-sso-as-mfa-method.md

```diff
@@ -0,0 +1,528 @@
+per-session MFA, moderated sessions, and MFA for admin actions. Using SSO as an
+MFA method opens up the possibility of poorly configured clusters being
+vulnerable to attacks ranging from internal users avoiding safe MFA checks to
+attackers with a compromised IdP gaining keys to the castle.
+
+#### Opt-in
+
+SSO as an MFA method will be opt-in. Administrators will be instructed through
+the docs to only enable an IdP connector as an MFA method if the IdP provider
+has strict checks itself (e.g. Administered Webauthn devices, Trusted devices).
+
+Teleport has no way to confirm whether a registered IdP connector follows the
+guidelines, but it will display a warning to admins who attempt to enable it.
+
+```console
+> tctl edit connector/okta
+### sets `allow_as_mfa_method: yes`
+Warning: Allowing this IdP provider to be used as an MFA method may reduce the
+security of enforced MFA checks for critical security features. This option
+should not be enabled unless the IdP provider has strict MFA and/or Device trust
+enforcement itself. Continue? (y/N):
+```
+
+#### IdP Compromise
+
+In the case of a full-scale IdP compromise, an attacker may have the ability
+to auto-provision users with arbitrary permissions.
+
+When device trust is required, newly auto-provisioned SSO users are required
+to add their first MFA device from a trusted device. When combined with MFA
+security features, such as MFA for Admin actions and per-session MFA, the blast
+radius of an IdP compromise is largely contained. The attacker would be
+prevented from accessing any critical infrastructure or making any changes
+to the cluster's security configuration.
+
+Allowing SSO as an MFA method would bypass the device trust check, opening the
+cluster back up to attacks in the case of an IdP compromise. To maintain this
+invariant, device trust must be enforced within the SSO MFA check.
+
+### Implementation
+
+#### Privilege Tokens
+
+Privileged tokens are used as transient MFA verification for some operations in
+Teleport today, like account resets. These tokens can be generated by passing a
+completed MFA challenge to `rpc CreatePrivilegeToken`.
+
+Currently, privilege tokens can only be used in select endpoints as a replacement
+for MFA. Privilege Tokens will be supported as a normal MFA response to support
+other use cases:
+
+```proto
+message MFAAuthenticateResponse {
+  oneof Response {
+    TOTPResponse TOTP = 2;
+    webauthn.CredentialAssertionResponse Webauthn = 3;
+    PrivilegeTokenResponse Token = 4;
+  }
+}
+```
+
+Note: Privilege tokens used in this way must be consumed upon verification to
+ensure that exchanging a webauthn or totp token for a privilege token through
+`rpc CreatePrivilegeToken` has no significant security implications.
+
+#### SSO MFA flow
+
+We will introduce a new SSO login flow that allows clients to get a new privilege
+token instead of the standard login credentials. Authenticated clients can create
```

Comment:
> We will introduce a new SSO login flow that allows clients to get a new privilege
token instead of the standard login credentials

Related to comments above, but why not use a distinct type altogether? A new type seems like less likely to open unwanted interactions.

Reply:
As in something separate from a privilege token? I felt that privilege token fits the bill well, but perhaps you're right.

Reply:
Sleep on it and let me know what you think.

Reply:
Agreed, I plan on adding a new user token type like `sso-mfa`.